CHANGELOG.md

```diff
@@ -48,6 +48,8 @@ balances or a single balance by denom when the `denom` query parameter is presen
 * [\#5785](https://github.com/cosmos/cosmos-sdk/issues/5785) JSON strings coerced to valid UTF-8 bytes at JSON marshalling time
 are now replaced by human-readable expressions. This change can potentially break compatibility with all those client side tools
 that parse log messages.
+* (codec) [\#5799](https://github.com/cosmos/cosmos-sdk/pull/5799) Now we favor the use of MarshalBinaryBare instead of LengthPrefixed in all cases that are not needed,
```

Comment:
This isn't true. It doesn't break CLI and/or REST commands or output (which are all JSON/text).

Reply:
true @alessio, what it breaks is the internal state, how some data is getting persisted but not how we talk with the application.

Reply:
From https://github.com/cosmos/gaia/pull/323/files#diff-ef04ec516e080b802253864b259addb3L1004

I may be wrong, so please correct me when I make a false statement:

```
success, base64Encoded, _ := f.TxEncode(jsonTxFile.Name())
```

- `TxEncode` runs the encode command, argument is filename. Output (string) is stored in `base64Encoded`. Output is subsequently trimmed and stored in `trimmedBase64` (https://github.com/cosmos/gaia/pull/323/files#diff-ef04ec516e080b802253864b259addb3L1006).

```
trimmedBase64 := strings.Trim(base64Encoded, "\"\n")
```

- Afterwards `trimmedBase64` goes through base64 decoding, output is stored in `decodedBytes`:

```
decodedBytes, err := base64.StdEncoding.DecodeString(trimmedBase64)
```

- We used to unmarshal `decodedBytes` through `UnmarshalBinaryLengthPrefixed`, now we use `UnmarshalBinaryBare`.

The last step has changed and it is due to the fact that internal encoding has changed, thus the client must use a different unmarshalling function (`UnmarshalBinaryBare` vs `UnmarshalBinaryLengthPrefixed`).

Let's stick to the ABI command-related point though: the command's output is produced and stored in `base64Encoded`. Given the same parameters and preconditions are passed, are those bytes produced by the encode command changed since #5748?

Reply:
No, client encoding has not changed.

Reply:
```
alessio@bangalter:~/work/gaia$ gaiacli version --long
{"name":"gaia","server_name":"gaiad","client_name":"gaiacli","version":"0.0.0-159-g680bb19","commit":"680bb196d6c5ece0c78875dff0175fdd780711ac","build_tags":"netgo,ledger","go":"go version go1.14 linux/amd64"}
alessio@bangalter:~/work/gaia$ gaiacli keys list
ERROR: unmarshal to keys.Info failed after 4 bytes (unrecognized prefix bytes 630DAD15): 630DAD153D0A03666F6F1226EB5AE987210363C8A9EFE5D7F200D99BAEB86AFEE26D93B6E032AC14740736B383918849F8AC1A25E1B0F79B20AEC5D05802A1B12F449BA6689B5759EBA7F3BB139D0F3640BFA8B37673DE8A6E2209736563703235366B31
```

(my keys were created, serialised and stored with `gaiacli` built from one of the most recent commits of `cosmos/gaia`)

Reply:
Test evidence follows:

```
alessio@bangalter:~/work/gaia$ git checkout 89e8df1882f6af6b939bc6f8a5f244b0cb04619d
Note: switching to '89e8df1882f6af6b939bc6f8a5f244b0cb04619d'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by switching back to a branch.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -c with the switch command. Example:

  git switch -c <new-branch-name>

Or undo this operation with:

  git switch -

Turn off this advice by setting config variable advice.detachedHead to false

HEAD is now at 89e8df1 Merge PR #312: proto: remove proto mentions
alessio@bangalter:~/work/gaia$ make install
go install -mod=readonly -tags "netgo ledger" -ldflags '-X github.com/cosmos/cosmos-sdk/version.Name=gaia -X github.com/cosmos/cosmos-sdk/version.ServerName=gaiad -X github.com/cosmos/cosmos-sdk/version.ClientName=gaiacli -X github.com/cosmos/cosmos-sdk/version.Version=0.0.0-158-g89e8df1 -X github.com/cosmos/cosmos-sdk/version.Commit=89e8df1882f6af6b939bc6f8a5f244b0cb04619d -X "github.com/cosmos/cosmos-sdk/version.BuildTags=netgo,ledger"' -trimpath ./cmd/gaiad
go install -mod=readonly -tags "netgo ledger" -ldflags '-X github.com/cosmos/cosmos-sdk/version.Name=gaia -X github.com/cosmos/cosmos-sdk/version.ServerName=gaiad -X github.com/cosmos/cosmos-sdk/version.ClientName=gaiacli -X github.com/cosmos/cosmos-sdk/version.Version=0.0.0-158-g89e8df1 -X github.com/cosmos/cosmos-sdk/version.Commit=89e8df1882f6af6b939bc6f8a5f244b0cb04619d -X "github.com/cosmos/cosmos-sdk/version.BuildTags=netgo,ledger"' -trimpath ./cmd/gaiacli
alessio@bangalter:~/work/gaia$ gaiacli keys list
[
  {
    "name": "foo",
    "type": "local",
    "address": "cosmos1n92t8cvq28p4ecy4cnedxkqtfyy7ff24tenpwy",
    "pubkey": "cosmospub1addwnpepqd3u3200uhtlyqxenwhts6h7ufke8dhqx2kpgaq8x6ec8yvgf8u2cm4kjv6"
  }
```

Reply:
(True, this has nothing to do with the `encode` command, but I stumbled upon this while I was trying to generate a sendTx transaction and feed it to the 2 different versions of the `encode` command)